### PR TITLE
[#2279] improvement(spark): Trigger the upstream rewrite when the read stage fails 

### PR DESCRIPTION
**Pull request by yl09099**
### What changes were proposed in this pull request?

If the current Reader fails to obtain Shuffle data, it does not trigger the upstream Stage to rewrite the data. If a Shuffle Server fails, it does not trigger Stage retry.

### Why are the changes needed?

Fix: https://github.com/apache/incubator-uniffle/issues/2279

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

**Comment by yl09099** [comment]
Wait a minute. I'm not ready for this.

**Comment by jerqi** [comment]
> Wait a minute. I'm not ready for this.

If this is not ready to review, you can mark this as draft.

**Comment by yl09099** [comment]
> > Wait a minute. I'm not ready for this.
> 
> If this is not ready to review, you can mark this as draft.

ok

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java [hunk 1]
```diff
@@ -553,16 +576,36 @@ public void incPartitionFetchFailure(int stageAttempt, int partition) {
           });
     }
 
-    public int getPartitionFetchFailureNum(int stageAttempt, int partition) {
+    public boolean getPartitionFetchFailureNum(
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:579]: This method has a weird return value.

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java [hunk 2]
```diff
@@ -553,16 +576,36 @@ public void incPartitionFetchFailure(int stageAttempt, int partition) {
...
+              if (this.partitions[partition] >= shuffleManager.getMaxFetchFailures()) {
+                return true;
+              } else {
+                return false;
+              }
             }
           });
     }
+
+    public void setClearedMapTrackerBlock(boolean isCleared) {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:595]: Could you have a better method name?

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java (hunk 1 above)
Comment (yl09099) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:579]: > This method has a weird return value.

The number of failures is compared to the maximum number of failures in the configuration, and I moved the comparison logic to the internal implementation, so I only need to return whether fetchfailed is thrown.

Comment (jerqi) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:579]: It's not proper name if we don't change the method name.

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java [hunk 3]
```diff
@@ -553,16 +576,36 @@ public void incPartitionFetchFailure(int stageAttempt, int partition) {
...
+              if (this.partitions[partition] >= shuffleManager.getMaxFetchFailures()) {
+                return true;
+              } else {
+                return false;
+              }
             }
           });
     }
+
+    public void clearedMapTrackerBlock(boolean isCleared) {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:595]: Maybe we don't the parameter `isCleared`. It's rebundant.

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java [hunk 4]
```diff
@@ -553,16 +576,36 @@ public void incPartitionFetchFailure(int stageAttempt, int partition) {
...
+      withWriteLock(
+          () -> {
+            this.isClearedMapTrackerBlock = isCleared;
+            return null;
+          });
+    }
+
+    public boolean isClearedMapTrackerBlock() {
```
Comment (jerqi) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:603]: `isClearedMapTrackerBlock` -> `hasClearedMapTrackerBlock`

Comment (yl09099) [reply client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java]: The above has been changed.

**Review by jerqi — APPROVED** [review]
LGTM.

**Comment by jerqi** [comment]
cc @maobaolong

**Pull request merged**